File: Year-2023/LeetCode-April-Challenge/ValidParentheses.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
private:
    bool isValidParenthesis(char a, char b){
        if(a == '(' && b == ')') return true;
        else if(a == '[' && b == ']') return true;
        else if(a == '{' && b == '}') return true;
        return false;
    }
public:
    bool isValid(string s) {
        stack<char> myStack; char checker;
        for(auto ch: s){
            if(ch == '(' || ch == '{' || ch == '[') myStack.push(ch);
            else if(ch == ')' || ch == '}' || ch == ']'){
                if(myStack.empty()) return false;
                checker = myStack.top();
                if(!isValidParenthesis(checker, ch)) return false;
                else myStack.pop();
            }
        }
        return myStack.empty() ? true: false;
    }
};
int main(){
    string s = "()[]{}";
    Solution *S = new Solution();
    cout<<std::boolalpha<<S->isValid(s)<<endl;
    return 0;
}

File: Year-2023/LeetCode-April-Challenge/ValidParentheses.js
var isValid = function (s) {
    const isValidParenthesis = (a, b) => {
        if (a === '(' && b === ')') return true;
        if (a === '{' && b === '}') return true;
        if (a === '[' && b === ']') return true;
        return false;
    }
    let stack = new Array(), checker = '';
    for (let ch of s) {
        if (ch === '(' || ch == '{' || ch == '[') stack.push(ch);
        else if (ch === ')' || ch === '}' || ch == ']') {
            if (stack.length === 0) return false;
            checker = stack.at(stack.length - 1);
            if (!isValidParenthesis(checker, ch)) return false;
            stack.pop();
        }
    }
    return stack.length === 0 ? true : false;
};
let s = "()[]{}";
console.log(isValid(s));